### ISSUE TITLE
Problems of deserializing nils and enums

### ISSUE THREAD

**Issue by jduan** [pinterest/riffed#31]
I've been playing with Riffed and I encountered a problem when writing tests.

My service is a simple CRUD service that allows clients to create, get, update, and delete animals. (think of the service as a zoo service)

In the test where I try to get an animal that doesn't exist, the assertion fails because of:

```
  1) test get animal that doesn't exist (ZooServerTest)
     test/zoo_server_test.exs:26
     Assertion with == failed
     code: response.error() == error
     lhs:  %ZooServer.Models.Error{errors: [%ZooServer.Models.ErrorEntry{message: "animal not found", location: nil, reason: 7}]}
     rhs:  %ZooServer.Models.Error{errors: [%ZooServer.Models.ErrorEntry{message: "animal not found", location: :undefined, reason: %ZooServer.Models.ErrorReason{ordinal: :resource_not_found, value: 7}}]}
     stacktrace:
       test/zoo_server_test.exs:32: (test)
```

So there are two problems:

1. The service doesn't specify a location so the client deserializes location as "nil". However, when I create an ErrorEntry manually on the client side, the location would default to ":undefined". The Riffed doc clearly says so but I wonder why struct fields won't default to nil?

2. The 2nd problem relates to  enums. When the client deserializes the reason, it sets it to 7. However, when I create an ErrorEntry manually, I use "Models.ErrorReason.resource_not_found" and I get a struct with ordinal and value. Obviously, you can't compare 7 with a struct object. So why won't Riffed or elixir-thrift deserializes 7 into a nice struct object? They should be able to do so based on number 7 I think.

**Comment by scohen** [comment]
Here's what I can think of:
1. the erlang version uses `:undefined` as its nil sentinel. Automatically converting from `:undefined` to `nil` could be dangerous, if your intention was to serialize `:undefined` in the first place. I know it's not ideal, but that's how it is. 

2. You need to tell riffed that you want your reason to be converted into a struct. You do this with the `enumerize_struct` macro.  There are some docs here: https://github.com/pinterest/riffed and 

It'll look like this:
```elixir
enumerize_struct ZooServer.Models.ErrorEntry, reason: ZooServer.Models.ErrorReason

# or, if you're already in the context of ZooServer.Models...
enumerize_struct ErrorEntry, reason: ErrorReason
```

**Comment by jduan** [comment]
@scohen thanks for your quick response. I fixed the code by doing the following:

1. I explicitly set "location" to nil when creating an object so Riffed won't use :undefined. That makes the equality assertion pass.

2. For the enum issue, I actually made a mistake myself. I should've  had used "reason" in the enumerize_struct directive but I used "type" instead.

For posterity, this is the commit I made: https://github.com/jduan/zoo_server/commit/53f2785737bd3254e7d3dff1ec0d365f31aa7305

Thanks again!
